OfflineAudioContext/explainer.md

```diff
@@ -0,0 +1,247 @@
+  const isRenderingCompleted = false;
+  while (!isRenderingCompleted) {
+    // get the next chunk of data from the stream
+    const result = await reader.read();
+
+    // the reader returns done = true when there are no more chunks to consume
+    if (result.done) {
+      break;
+    }
+
+    // result.value contains interleaved Float32Array values
+    const buffers = result.value;
+  }
+} else {
+  audioBuffer = await offlineAudioContext.startRendering();
+}
+```
+
+Proposed interface:
+
+```js
+// From https://developer.mozilla.org/en-US/docs/Web/API/AudioData/format
+enum AudioFormat {
+    "u8",
+    "s16",
+    "s32",
+    "f32",
+    "u8-planar",
+    "s16-planar",
+    "s32-planar",
+    "f32-planar"
+    "f32-interleaved"
+}
+
+dictionary OfflineAudioRenderingOptions {
+    required AudioFormat format = "f32-interleaved";
+}
+
+partial interface OfflineAudioContext {
+    // Returns a stream that yields buffers of interleaved audio samples in Float32Array or whatever format is specified
+    Promise<ReadableStream> startRenderingStream(options?: OfflineAudioRenderingOptions);
+};
+```
+
+### Pros
+
+- The new capability is feature detectable because it is a new function. Compared to Alternative 1 which cannot be easily detected
+- Aligns well with other web streaming APIs, similar to [WebCodecs](https://streams.spec.whatwg.org/#readablestream)
+- Works with very large durations, no upper limit to WebAudio graph duration
+
+### Cons
+
+- Need to define sensible chunk sizes, backpressure, error handling, and end-of-stream
+
+### Output format
+
+There is an open question of what data format `startRenderingStream()` should return. The options under consideration are `AudioBuffer`, `Float32Array` planar or `Float32Array` interleaved.
+
+#### `AudioBuffer`
+
+**Pros**
+
+- semantically closest to the `startRendering()` API
+- does not add a new type to the WebAudio spec
+
+**Cons**
+
+- does not allow developers to BYOB (bring your own buffer) and BYOB helps developers manage memory usage, so `AudioBuffer` removes a bit of control
+
+#### `f32-planar` per channel
+
+**Pros**
+
+- `f32-planar` also already exists in the WebAudio spec
+
+**Cons**
+
+- requires the output of `startStreamingRendering()` to return an array of `Float32Array` in planar format for each output channel
+- this leaves a question of what to do if only one channel is read by the consumer, i.e. what should happen to the other channel's data?
+
+#### `f32-interleaved`
+
+**Pros**
+
+- allows for streaming a single stream of data, rather than one for each channel
+- interoperates with WebCodecs APIs because they operate with interleaved streams of data
+
+**Cons**
+
+- introduces a new type to the WebAudio spec, `f32-interleave` which does not exist at the moment
+
+#### Recommendation
+
+`f32-interleaved` as it is the most interoperable with other media APIs, like WebCodecs, and simplifies processing with other data streams such as video.
```

Comment:
Again, WebCodecs supports both. Idk what other web APIs there are that only support interleaved?

Reply:
So what should we put here for the format? `f32` or `f32-planar`?